src/systems/hydrodynamics/Hydrodynamics.cc

```diff
@@ -78,14 +79,121 @@ class gz::sim::systems::HydrodynamicsPrivateData
   /// \brief Previous state.
   public: Eigen::VectorXd prevState;
 
+  /// \brief Prefious derivative of the state
   public: Eigen::VectorXd prevStateDot;
 
-  /// Link entity
+  /// \brief Use current table if true
+  public: bool useCurrentTable {false};
+
+  /// \brief Current table xComponent
+  public: std::string axisComponents[3];
+
+  public: std::shared_ptr<gz::sim::v7::components::EnvironmentalData>
+    gridField[3];
+
+  public: std::optional<gz::math::InMemorySession<double, double>> session[3];
+
+  /// \brief Link entity
   public: Entity linkEntity;
 
+  /// \brief has dat been loaded
+  public: bool ready{false};
+
   /// \brief Ocean current callback
   public: void UpdateCurrent(const msgs::Vector3d &_msg);
 
+  /// \brief Set the current table
+  public: void SetCurrentTable(
+    const EntityComponentManager &_ecm,
+    const std::chrono::steady_clock::duration &_currTime)
+  {
+    _ecm.EachNew<components::Environment>([&](const Entity &/*_entity*/,
+    const components::Environment *_environmental_data)->bool
+    {
+      for (std::size_t i = 0; i < 3; i++)
+      {
+        if (this->axisComponents[i] != "")
+        {
+          auto data = _environmental_data->Data();
+          if (!data->frame.Has(this->axisComponents[i]))
+          {
+            gzwarn << "Environmental sensor could not find field "
+              << this->axisComponents[i] << "\n";
+            continue;
+          }
+
+          this->gridField[i] = data;
+          this->session[i] =
+            this->gridField[i]->frame[this->axisComponents[i]].CreateSession();
+          if (!this->gridField[i]->staticTime)
+          {
+            this->session[i] =
+              this->gridField[i]->frame[this->axisComponents[i]].StepTo(
+                *this->session[i],
+                std::chrono::duration<double>(_currTime).count());
+          }
+          this->ready = true;
+
+          if(!this->session[i].has_value())
+          {
+            gzerr << "Exceeded time stamp." << std::endl;
+          }
+        }
+      }
+      return true;
+    });
+  }
+
+  /// Retrieve the currernt from the data table
+  public: math::Vector3d GetCurrentFromEnvironment(
+    const EntityComponentManager &_ecm,
+    const std::chrono::steady_clock::duration &_currTime,
+    math::Vector3d _position)
+  {
+    math::Vector3d current(0, 0, 0);
+
+    if (!this->ready)
```

Comment:
@arjo129 nit: isn't `ready` equivalent to having (1) a data frame and (b) at least one session for some axis? Do we need the (redundant) boolean flag?